File: tests/test_bv.py
"""
Test BV invariant -- Sage
Run: sage -python tests/test_bv.py [test_data_file]   (from repo root)

Uses PARI's qfminim convention (v^T * M * v <= d) to match test_bv.gp and test_bv.m.
Defaults to tests/test_bv_data.txt if no file is specified.
"""
import sys, os, time
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'alt'))
from bv import BV, HBV_poly, HBV_xor
from sage.all import Matrix, ZZ

def load_test_data(path):
    """Parse test_bv_data.txt -> list of (gram, expected_poly, expected_xor)."""
    data = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            gram_str, poly_str, xor_str = line.split(':')
            # parse [a,b,...] -> list of ints
            entries = list(map(int, gram_str.strip().strip('[]').split(',')))
            n = int(len(entries)**0.5)
            assert n * n == len(entries), f"entry count {len(entries)} is not a perfect square"
            gram = Matrix(ZZ, n, n, entries)
            data.append((gram, int(poly_str.strip()), int(xor_str.strip())))
    return data

D = 4  # short vector bound
if len(sys.argv) > 1:
    data_path = sys.argv[1]
else:
    data_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'test_bv_data.txt')
test_data = load_test_data(data_path)

print("=" * 60)
print("BV invariant test -- Sage")
print("=" * 60)

t0 = time.time()
total_bv = total_poly = total_xor = 0
ok = True
for i, (gram, exp_poly, exp_xor) in enumerate(test_data):
    assert gram == gram.T, f"Matrix {i+1} is not symmetric"
    t1 = time.time()
    bv = BV(gram, D)
    t2 = time.time()
    hp = HBV_poly(bv)
    t3 = time.time()
    hx = HBV_xor(bv)
    t4 = time.time()
    total_bv += t2 - t1
    total_poly += t3 - t2
    total_xor += t4 - t3
    print(f"  Matrix {i+1} ({gram.nrows()}x{gram.ncols()}): poly = {hp}  xor = {hx}  (BV {t2-t1:.4f}s  poly {t3-t2:.4f}s  xor {t4-t3:.4f}s)")
    if hp != exp_poly:
        print(f"FAIL: matrix {i+1} poly hash mismatch: got {hp}, expected {exp_poly}")
        ok = False
    if hx != exp_xor:
        print(f"FAIL: matrix {i+1} xor hash mismatch: got {hx}, expected {exp_xor}")
        ok = False
print(f"  Total: {time.time() - t0:.3f}s  (BV {total_bv:.3f}s  poly {total_poly:.4f}s  xor {total_xor:.4f}s)")

if ok:
    print("PASS: all hashes match expected values (cross-implementation verified)")

File: tests/test_bv.m
/*
 * Test BV invariant -- Magma
 * Run: magma -b tests/test_bv.m   (from repo root)
 *
 * Test data read from tests/test_bv_data.txt.
 */

SetColumns(0);

load "alt/bv.m";

function LoadTestData(path)
    data := [* *];
    F := Open(path, "r");
    while true do
        line := Gets(F);
        if IsEof(line) then break; end if;
        line := StripWhiteSpace(line);
        if #line eq 0 or line[1] eq "#" then continue; end if;
        parts := Split(line, ":");
        gram_str := StripWhiteSpace(parts[1]);
        // strip [ and ]
        gram_str := gram_str[2..#gram_str-1];
        entries := [StringToInteger(StripWhiteSpace(x)) : x in Split(gram_str, ",")];
        n := Isqrt(#entries);
        assert n^2 eq #entries;
        gram := Matrix(Integers(), n, n, entries);
        poly_h := StringToInteger(StripWhiteSpace(parts[2]));
        xor_h := StringToInteger(StripWhiteSpace(parts[3]));
        Append(~data, <gram, poly_h, xor_h>);
    end while;
    return data;
end function;

// ---- Tests ----

D := 4;
data := LoadTestData("tests/test_bv_data.txt");

printf "============================================================\n";
printf "BV invariant test -- Magma\n";
printf "============================================================\n";

t0 := Cputime();
tot_bv := 0.0; tot_poly := 0.0; tot_xor := 0.0;
ok := true;
for i := 1 to #data do
    gram := data[i][1];
    exp_poly := data[i][2];
    exp_xor := data[i][3];
    n := Nrows(gram);
    assert gram eq Transpose(gram);
    t1 := Cputime();
    bv := BV(gram, D);
    t2 := Cputime();
    hp := HBV_poly(bv);
    t3 := Cputime();
    hx := HBV_xor(bv);
    t4 := Cputime();
    tot_bv +:= t2 - t1; tot_poly +:= t3 - t2; tot_xor +:= t4 - t3;
    printf "  Matrix %o (%ox%o): poly = %o  xor = %o  (BV %os  poly %os  xor %os)\n",
           i, n, n, hp, hx, t2-t1, t3-t2, t4-t3;
    if hp ne exp_poly then
        printf "FAIL: matrix %o poly hash mismatch: got %o, expected %o\n", i, hp, exp_poly;
        ok := false;
    end if;
    if hx ne exp_xor then
        printf "FAIL: matrix %o xor hash mismatch: got %o, expected %o\n", i, hx, exp_xor;
        ok := false;
    end if;
end for;
printf "  Total: %os  (BV %os  poly %os  xor %os)\n", Cputime(t0), tot_bv, tot_poly, tot_xor;

if ok then
    printf "PASS: all hashes match expected values (cross-implementation verified)\n";
end if;

quit;
